# Mathematical Operators 

## Prerequisites

- Variables and their data types
- Printing f-strings

## Learning outcomes

- Develop familiarity with basic mathematical operations in Python
- Understand how to access some additional, more complex mathematical operations. 

## Arithmetic

Python is extremely relevant to numerical computing, in particular, thanks to the presence of the Python library [NumPy](ADD A LINK TO THE NUMPY LIBRARY PAGE). 
Therefore, it is useful to outline some of the mathematical operations that can be natively performed with Python. 

Python natively supports some mathematical operations. 

| Operation    | Mathematical Notation | Pythonic Notation |
| -------- | ------- | ------- |
| Addition  | $a + b$ | `a + b` |
| Subtraction | $a - b$ | `a - b` |
| Multiplication | $a \times b$ | `a * b` |
| Division | $a \div b$ | `a / b` |
| Exponent | $a ^ b$ | `a ** b` |
| Modulo | $a \textrm{ mod } b$ | `a % b` |

The modulo operation may be new to you, you may know it as the remainder from the division of two numbers. 

As we saw in the example above, a single line of code may have many mathematical operations. 
In this event, Python will follow the standard order of operations for mathematical operations: you make know this as [BODMAS](https://en.wikipedia.org/wiki/Order_of_operations#Mnemonics).

### Example: The Quadratic Formula

The quadratic formula is an expression to solve quadratic equations with the form, 

$$
ax^2 + bx + c = 0,
$$

where, $x$ is an unknown value that we want to find, and $a$, $b$, and $c$ are fixed parameters. 
The quadratic formula states that the value of $x$ is, 

$$
x = \frac{-b \pm \sqrt{b^2 - 4ac}}{2a}.
$$

This becomes relevant in chemistry if we consider the following problem: Formic acid is a weak acid with a dissociation constant $K_a$ of 1.8&times;10<sup>-4</sup>. 
The dissociation constant relates the concentration of the H<sup>+</sup> ions and the amount of acid dissolve, $N$, by the equation: 

$$
K_a = \frac{[\textrm{H}^+]^2}{N - [\textrm{H}^+]}.
$$

This equation can be reformulated as a quadratic equation, 

$$
[\textrm{H}^+]^2 + K_a[\textrm{H}^+] - K_aN = 0.
$$

Therefore, we can use the quadratic formula to solve for the concentration of hydrogen ions for a given amount of dissolved acid, where $a = 1$, $b=K_a$ and $c=-K_aN$. 

We can write Python code to compute [H<sup>+</sup>] for 0.1 moles of dissolved acid. 
Don't worry if some of the code specifics below are a bit new, the table below explains each of the operators. 

In [1]:
K_a = 1.8e-4
N = 0.1

a = 1.
b = K_a
c = -K_a * N

H_conc_plus = (-b + (b ** 2 - 4 * a * c) ** (1 / 2)) / (2 * a)
H_conc_minus = (-b - (b ** 2 - 4 * a * c) ** (1 / 2)) / (2 * a)

print(f'H_conc_plus = {H_conc_plus:.5f} M')
print(f'H_conc_minus = {H_conc_minus:.5f} M')

H_conc_plus = 0.00415 M
H_conc_minus = -0.00433 M


Clearly, the variable `H_conc_minus` should be disregarded as it is not possible to have a negative concentration. 
This means that the concentration of [H<sup>+</sup>] is 0.00415 (to 5 decimal places). 

## More Complex Mathematical Operations

The `numpy` module provides access to a range of advanced mathematical functions. 
Information about all of the functions that the `numpy` module has can be found in the [numpy lesson](LINK TO NUMPY LESSON). 
To access a given function, we must *import* it *from* the module. 
Below, we import the base 10 logarithm function, `log10`. 

In [3]:
from numpy import log10

We can then use this with the result from above to compute the pH of the solution. 

In [4]:
pH = log10(H_conc_plus)
print(f'pH = {pH:.2f}')

pH = -2.38


### Exercise
Find the velocity, $v$, of a N<sub>2</sub> molecule with a mass, $m$ of 4.6&times;10<sup>-26</sup> kg at a temperature, $T$, of 293 K, given the following equation,

$$
v = \sqrt{\frac{3k_bT}{m}}, 
$$

where, $k_b$ is 1.38&times;10<sup>−23</sup> J/K.

#### Answer

In [2]:
k_b = 1.38e-23
T = 293
m = 4.6e-26

v = ((3 * k_b * T) / m) ** (1 / 2)
print(f'velocity = {v:.1f} m/s')

velocity = 513.5 m/s


## Learning outcomes

- Develop familiarity with basic mathematical operations in Python, including `+`, `-`, `*`, `/`. 
- Understand how to access some additiona, more complex mathematical function using the `numpy` module. 

# TODO 

- Add more exercises for other operations